src/transformers/modeling_xlnet.py

```diff
@@ -1000,7 +1000,15 @@ def prepare_inputs_for_generation(self, input_ids, past, **kwargs):
 
         effective_batch_size = input_ids.shape[0]
         dummy_token = torch.zeros((effective_batch_size, 1), dtype=torch.long, device=input_ids.device)
-        input_ids = torch.cat([input_ids, dummy_token], dim=1)
+
+        # at every pass, the attention values for the new token and the two last generated tokens
+        # are computed, the rest is reloaded from the `past` cache
+        offset = 2
```

Comment:
I looked a bit deeper into the code, especially this line:
https://github.com/huggingface/transformers/blob/b25f7802de2749a5f8c3430437eceabf9e6384b8/src/transformers/modeling_xlnet.py#L336


My conclusion is 
a) 
key, value are always recalculated for each memory position -> so that there is actually 0 logic in the code that speeds up attention. We also saw this in our timing experiments. To me it seems that key and value projections are recalculated for each past memory position even though this position was already calculated before and thanks to causal masking should not influence future calculations.
b) 
mems are completely different do `past` as it is in GPT2 and are *not* key, value projections, but hidden states (the outputs of attention layers) => so we might need to add a complete new `past_key_value` varable here to speed up inference. Inference is extremely slow in XLNet and we had many issues about it. Looking at the code, I think we can implement this and could gain significantly speed ups *with equal* mathematical output (not a 100% sure though - will have to dive deeper in the code...what do you think @TevenLeScao ?)
c)
The code as implemented by you and as discussed I think is logically the correct way to implement causal language modelling for XLNet, but setting `offset = 2` seems a bit random here...maybe we can even contact the author for more information how exactly causal lm should be implemented as setting offest just to 2 seems a bit random to me.